### ISSUE TITLE
tautological warnings for is_constant_evaluated not emitted in c++23/26 modes

### ISSUE THREAD

**Issue by cor3ntin** [llvm/llvm-project#125955]
for

```cpp

int main() {
   if constexpr(std::is_constant_evaluated()) { }
}
```
We usually warn 
> 'std::is_constant_evaluated' will always evaluate to 'true' in a manifestly constant-evaluated expression [-Wconstant-evaluated]

However this warning is gone in c++23 mode

https://godbolt.org/z/KKPKvveGM

**Comment by jwakely** [comment]
> Report an issue against libstdc++?

What would the issue be? Thate use a more efficient implementation than the built-in?

In C++23 libstdc++ defines `std::is_constant_evaluated` using `if consteval { return true; } else { return false; }` because it's cheaper for GCC to compiler that.

```c++
  [[__gnu__::__always_inline__]]
  constexpr bool
  is_constant_evaluated() noexcept
  {
#if __cpp_if_consteval >= 202106L
    if consteval { return true; } else { return false; }
#else
    return __builtin_is_constant_evaluated();
#endif
  }
```

In theory we could change that to `#if __cpp_if_consteval >= 202106L && ! __clang__` but I'd rather not.

**Comment by jwakely** [comment]
Reduced:

```c++
namespace std
{
    [[gnu::always_inline]]
    constexpr bool is_constant_evaluated()
    {
#if __cpp_if_consteval >= 202106L
        if consteval { return true; }
        else { return false; }
#else
        return __builtin_is_constant_evaluated();
#endif
    }
}

int main() {
   if constexpr(std::is_constant_evaluated()) { }
}
```

https://godbolt.org/z/PPd6vs7TK

**Comment by tbaederr** [comment]
Those warnings are emitted from https://github.com/llvm/llvm-project/blob/cf444ac2adc45c1079856087b8ba9a04466f78db/clang/lib/AST/ExprConstant.cpp#L13661-L13674

We could of course do a similar check for general calls and check if those calls are to `std::is_constant_evaluated()`, but that would be even uglier.

**Comment by localspook** [comment]
> In theory we could change that to #if __cpp_if_consteval >= 202106L && ! __clang__ but I'd rather not.

That’s what I had in mind, and it’s a simpler change than fixing it in clang, so I threw the idea out there. I agree that making changes in libstdc++ specifically to accommodate clang isn’t ideal.
